lib/web/sshlogin.go

```diff
@@ -196,6 +201,89 @@ func SSHAgentLogin(proxyAddr, user, password, hotpToken string, pubKey []byte, t
 	return out, nil
 }
 
+// SSHAgentU2fLogin requests a U2F sign request (authentication challenge) via the proxy.
+// If the credentials are valid, the proxy wiil return a challenge.
+// We then call the official u2f-host binary to perform the signing and pass the signature to the proxy.
+// If the authentication succeeds, we will get a temporary certificate back
+func SSHAgentU2fLogin(proxyAddr, user, password string, pubKey []byte, ttl time.Duration, insecure bool, pool *x509.CertPool) (*SSHLoginResponse, error) {
+	clt, _, err := initClient(proxyAddr, insecure, pool)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	u2fSignRequest, err := clt.PostJSON(clt.Endpoint("webapi", "u2f", "signrequest"), u2fSignRequestReq{
+		User: user,
+		Pass: password,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Pass the JSON-encoded data undecoded to the u2f-host binary
+	u2fFacet := "https://" + strings.ToLower(proxyAddr)
+	cmd := exec.Command("u2f-host", "-aauthenticate", "-o", u2fFacet)
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cmd.Start()
+	stdin.Write(u2fSignRequest.Bytes())
+	stdin.Close()
+	fmt.Println("Please press the button on your U2F key")
+
+	// The origin URL is passed back base64-encoded and the keyHandle is passed back as is.
+	// A very long proxy hostname or keyHandle can overflow a fixed-size buffer.
+	signResponseLen := 500 + len(u2fSignRequest.Bytes()) + len(proxyAddr) * 4 / 3
+	signResponseBuf := make([]byte, signResponseLen)
+	signResponseLen, err = io.ReadFull(stdout, signResponseBuf)
+	// unexpected EOF means we have read the data completely.
```

Comment:
according to the docs of  https://golang.org/pkg/io/#ReadFull

```
ReadFull reads exactly len(buf) bytes from r into buf. It returns the number of bytes copied and an error if fewer bytes were read. The error is EOF only if no bytes were read. If an EOF happens after reading some but not all the bytes, ReadFull returns ErrUnexpectedEOF. On return, n == len(buf) if and only if err == nil.
```

so `EOF` means no data, `ErrUnexpectedEOF` means not all data read, and `err==nil` means we have just fit in, so if I read this right we need to check for `EOF` 

```go
if err != nil || err != io.ErrUnexpectedEOF {
   return nil, trace.LimitExceeded("u2f sign response exceeded buffer size")
}
```